### ISSUE TITLE
SherpaNcnn own model 

### ISSUE THREAD

**Issue by HassanTen** [k2-fsa/sherpa-ncnn#368]
Can I create my own model using Google Colab or another method without using the local machine, which fits with an Android application using Kotlin (SherpaNcnn)? Please clarify the steps.

**Comment by csukuangfj** [comment]
please describe in detail the difficulty you have. Would be.great if you.also.post error logs.

**Comment by HassanTen** [comment]
H

**Comment by HassanTen** [comment]
1. How can I convert a Wav2Vec 2.0 model to SherpaNcnn on Google Colab?
2. How can I create a custom SherpaNcnn model on Google Colab?
I am a new developer in creating custom models. I have searched on ChatGPT, and it says you will face difficulty.

**Comment by HassanTen** [comment]
> please describe in detail the difficulty you have. Would be.great if you.also.post error logs.

I couldn't locate the file. 
Can you share the link to LSTM transducer models on Google Colab?"**

**Comment by csukuangfj** [comment]
> How can I  xxx on Google Colab

Sorry, not possible.

**Comment by HassanTen** [comment]
> > How can I  xxx on Google Colab
> 
> Sorry, not possible.

I couldn't locate the file. 
Can you share the link to  streaming Zipformer transducer models on Google Colab?

**Comment by csukuangfj** [comment]
> Can you share the link to streaming Zipformer transducer models on Google Colab?



Where do you get the info that it exists?

**Comment by HassanTen** [comment]
> > Can you share the link to streaming Zipformer transducer models on Google Colab?
> 
> Where do you get the info that it exists?



Do you think these steps will succeed?

Here's the fully translated English version of the Google Colab notebook for training a Zipformer Transducer model on the Yes/No dataset using Icefall, with accuracy evaluation and export to Sherpa-NCNN:

---

🔁 1. Environment Setup

!pip install -q git+https://github.com/k2-fsa/icefall.git
!pip install -q k2 lhotse torchaudio pnnx sherpa-ncnn


---

📦 2. Download and Prepare the Yes/No Dataset

import os
from lhotse.recipes import prepare_yesno

os.makedirs("data/manifests", exist_ok=True)
manifests = prepare_yesno("data/manifests")
print("✅ Yes/No dataset is ready.")


---

🧠 3. Define the Zipformer + Transducer Model

import torch
import torch.nn as nn
from icefall.models.zipformer.zipformer import StreamingZipformer
from icefall.models.transducer.transducer import TransducerModel

# Define vocabulary
tokens = ["<blk>", "yes", "no", "<eos>"]
token_dict = {t: i for i, t in enumerate(tokens)}

# Define the model
class YesNoModel(TransducerModel):
    def __init__(self):
        encoder = StreamingZipformer(
            input_dim=80,
            output_dim=256,
            encoder_dim=256,
            proj_dim=128,
            num_encoder_layers=2,
            short_chunk_size=4,
            segment_size=8,
            left_context=16,
            right_context=0
        )
        decoder_dim = 128
        joiner_dim = 128

        super().__init__(
            encoder=encoder,
            decoder_dim=decoder_dim,
            num_decoder_tokens=len(tokens),
            joiner_dim=joiner_dim
        )

model = YesNoModel()
print("✅ Model is ready.")


---

🎧 4. Feature Extraction and Data Preparation

from lhotse.dataset import SimpleCut, CutSet
from lhotse import Fbank, FbankConfig
from torch.utils.data import DataLoader

fbank = Fbank(FbankConfig(num_mel_bins=80))
cuts = CutSet.from_manifests(**manifests).compute_and_store_features(fbank, "data/fbank")

train_cuts = cuts.subset(first=40)
test_cuts = cuts.subset(last=20)

def collate_fn(cuts):
    feats = [torch.tensor(cut.load_features()) for cut in cuts]
    texts = [cut.supervisions[0].text.lower() for cut in cuts]
    tokens_batch = [[token_dict["<blk>"]] + [token_dict[t]] + [token_dict["<eos>"]] for t in texts]
    return feats, tokens_batch, texts

train_dl = DataLoader(train_cuts, batch_size=4, collate_fn=collate_fn)
test_dl = DataLoader(test_cuts, batch_size=2, collate_fn=collate_fn)


---

🚀 5. Training with RNNT Loss

from icefall.training.rnnt_loss import compute_rnnt_loss

optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
model.train()

for epoch in range(5):
    for feats, token_ids, _ in train_dl:
        feats = nn.utils.rnn.pad_sequence(feats, batch_first=True)
        token_ids = nn.utils.rnn.pad_sequence(
            [torch.tensor(t) for t in token_ids], batch_first=True)

        loss = compute_rnnt_loss(model, feats, token_ids)
        loss.backward(); optimizer.step(); optimizer.zero_grad()

    print(f"✅ Epoch {epoch+1} complete, Loss = {loss.item():.4f}")


---

🧪 6. Evaluation and Accuracy

model.eval()
correct = total = 0

with torch.no_grad():
    for feats, token_ids, texts in test_dl:
        feats = nn.utils.rnn.pad_sequence(feats, batch_first=True)
        hyps = model.decode(feats)

        for ref, hyp in zip(texts, hyps):
            print(f"🔹 Reference: {ref} | Prediction: {tokens[hyp[0]]}")
            if tokens[hyp[0]] == ref:
                correct += 1
            total += 1

acc = 100 * correct / total
print(f"✅ Accuracy: {acc:.2f}%")


---

🧳 7. Export to Sherpa-NCNN

# Save model
os.makedirs("exp", exist_ok=True)
torch.save(model.state_dict(), "exp/yesno-model.pt")

# Clone Sherpa-NCNN and export script
!git clone https://github.com/k2-fsa/sherpa-ncnn
!cp -r sherpa-ncnn/scripts/icefall/export-for-ncnn.py .

# Export
!python3 export-for-ncnn.py \
  --tokens '["<blk>", "yes", "no", "<eos>"]' \
  --checkpoint exp/yesno-model.pt \
  --output-dir exported_model_yesno

**Comment by csukuangfj** [comment]
Sorry, you need to debug your own code.

If you have issues with our code examples, we can help you.
